server/src/main/scala/services/crunch/deskrecs/RunnableDeskRecs.scala

```diff
@@ -0,0 +1,125 @@
+
+    val crunchPeriodStartMillis: SDateLike => SDateLike = crunchStartWithOffset(airportConfig.crunchOffsetMinutes)
+
+    val graph = GraphDSL.create(
+      Source.actorRefWithAck[List[Long]](Ack).async,
+      KillSwitches.single[DeskRecMinutes])((_, _)) {
+      implicit builder =>
+        (daysToCrunchAsync, killSwitch) =>
+          val deskRecsSink = builder.add(Sink.actorRefWithAck(portStateActor, StreamInitialized, Ack, StreamCompleted, StreamFailure))
+          val parallelismLevel = 2
+
+          daysToCrunchAsync.out
+            .map(_.map(min => crunchPeriodStartMillis(SDate(min)).millisSinceEpoch))
+            .statefulMapConcat {
+              processQueue(parallelismLevel)
+            }
+            .mapAsync(parallelismLevel) { crunchStartMillis =>
+              log.info(s"Asking for flights for ${SDate(crunchStartMillis).toISOString()}")
+              flightsToCrunch(minutesToCrunch, askablePortStateActor, crunchStartMillis)
+            }
+            .map { case (crunchStartMillis, flights) =>
+              log.info(s"Crunching ${SDate(crunchStartMillis).toISOString()} flights: ${flights.flightsToUpdate.size}")
+              crunchFlights(flights, crunchStartMillis, minutesToCrunch, crunch, airportConfig)
+            }
+            .map(drms => DeskRecMinutes(drms.values.toSeq)) ~> killSwitch ~> deskRecsSink
+
+          ClosedShape
+    }
+
+    RunnableGraph.fromGraph(graph)
+  }
+
+  private def crunchFlights(flights: FlightsWithSplits,
+                            crunchStartMillis: MillisSinceEpoch,
+                            minutesToCrunch: Int,
+                            crunch: TryCrunch,
+                            airportConfig: AirportConfig): SortedMap[TQM, CrunchApi.DeskRecMinute] = {
+    val crunchEndMillis = SDate(crunchStartMillis).addMinutes(minutesToCrunch).millisSinceEpoch
+    val terminals = flights.flightsToUpdate.map(_.apiFlight.Terminal).toSet
+    val loadMinutes = WorkloadCalculator.flightLoadMinutes(flights, airportConfig.defaultProcessingTimes).minutes
+
+    val loadsWithDiverts = loadMinutes
+      .groupBy {
+        case (TQM(t, q, m), _) => val finalQueueName = airportConfig.divertedQueues.getOrElse(q, q)
+          TQM(t, finalQueueName, m)
+      }
+      .map {
+        case (tqm, mins) =>
+          val loads = mins.values
+          (tqm, LoadMinute(tqm.terminalName, tqm.queueName, loads.map(_.paxLoad).sum, loads.map(_.workLoad).sum, tqm.minute))
+      }
+
+    crunchLoads(loadsWithDiverts, crunchStartMillis, crunchEndMillis, terminals, airportConfig, crunch)
+  }
+
+  private def processQueue(parallelismLevel: Int): () => List[MillisSinceEpoch] => List[MillisSinceEpoch] = {
+    () =>
+      var queue = SortedSet[MillisSinceEpoch]()
+      incoming => {
+        queue = queue ++ incoming
+        val nextToProcess = queue match {
+          case q if q.nonEmpty =>
+            val head = q.take(parallelismLevel)
```

Comment:
rename head to something like next to process

Reply:
Done